FileStatusList: Add toolbar and actions on folders

mstv opened the pull request:
## Proposed changes

`RevisionDiffControl` if folder selected:
- Display names of contained items (with relative paths)
- Enable actions on contained items

`FileStatusList`:
- Add toolbar
- Align git-grep and filter boxes

## Screenshots <!-- Remove this section if PR does not change UI -->

### Before

![image](https://github.com/user-attachments/assets/1a0be075-2c31-44dd-91bc-8cd76acbfc33)

![image](https://github.com/user-attachments/assets/326c75ca-04f8-43ab-bc8a-b30e18160e4d)

### After

![image](https://github.com/user-attachments/assets/be029b54-8a7a-49d2-8105-87bd9f9fcfcf)

![image](https://github.com/user-attachments/assets/3fe67693-0bbc-40fe-95c0-ebc7e91d90a3)

![image](https://github.com/user-attachments/assets/94596cfa-7812-40fb-8fe8-4b61688c4058)

## Test methodology <!-- How did you ensure quality? -->

- manual
- extended tests

## Merge strategy

<!-- Change the following if the merge strategy should be changed:
- Squash merge (maintainer to decide merge message, PR submitter should cleanup commits/messages at PR approval).
- Rebase merge (PR submitter must change the commit message for the last commit).
- Merge commit. (PR submitter to rebase and squash before merges).
- To be decided later.
The maintainer may still request the contributor to squash and rebase, to make sure that merges and commit messages are clarified.
-->

I agree that the maintainer squash merge this PR (if the commit message is clear).

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

gerhardol commented:
A few comments on usage:
* Tooltip covers the text
![image](https://github.com/user-attachments/assets/2e15fa32-b800-4001-9c91-deaa7b1f7d84)
* When is Dense tree/Group nodes not disabled? Is dropdown needed?
* In some situations the list is not updated when clicking Group and/or !AB=. This occurs when selecting 421d22f65fdb8344544d011be1885f14e753223b and some situation with commits in two branches.

Noted, probably OK:
* If a folder does not exist on disk "Show in folder" is disabled, as files are. Good. If top level group or status/extension status is shown it is available, opening the repo. Maybe change text to "Show repo folder"?
* (maybe not this PR) If selecting a folder, "Save as" and others are available. This should maybe select the files in the folder too.

Some enhancements, probably future PRs:
* Add File tree group, similar to the grep group. The file tree is improved in an update, but then it would not be needed.
* Refresh button, for artificial commits. That can later show the status from GitStatusMonitor if the contents have changed.

mstv commented:
> Tooltip covers the text

Yes, not nice. It disappears on mouse move. Closing the dropdown does not work either. Blame WF.
Solution: Use `ToolStripEx` as in FormBrowse.

> When is Dense tree/Group nodes not disabled? Is dropdown needed?

- not implemented
- when someone insists on
  * having a separate folder node for single-file folders
  * omitting group-by nodes (perhaps, only for flat list)
 
> In some situations the list is not updated when clicking Group and/or !AB=.
> This occurs when selecting [421d22f](https://github.com/gitextensions/gitextensions/commit/421d22f65fdb8344544d011be1885f14e753223b) and some situation with commits in two branches.

If there are _modified_ `.cs` files only, there is just one group - and hence no difference because singular group-by folder nodes are omitted.

> Noted, probably OK:
> 
> If a folder does not exist on disk "Show in folder" is disabled, as files are. Good. If top level group or status/extension status is shown it is available, opening the repo. Maybe change text to "Show repo folder"?

I do not like to exchange menu item texts on runtime unless really necessary. I think opening the repo folder for root / group-by nodes is not counter-intuitive. 

> (maybe not this PR) If selecting a folder, "Save as" and others are available. This should maybe select the files in the folder too.

If you mean "Show in folder": A folder is selected. So I think, that this folder shall be selected in the Explorer.
Selecting multiple files opens multiple Explorer windows. If users really want all files to be shown in Explorer (windows), they should explicitly use "Select all" first.

> Some enhancements, probably future PRs:
> 
> Add File tree group, similar to the grep group. The file tree is improved in an update, but then it would not be needed.

File tree needs slightly different handling. E.g. grouping does not really apply. I prefer to not add more groups, especially no expensive ones. It would need to be excluded from "find text in files".

> Refresh button, for artificial commits. That can later show the status from GitStatusMonitor if the contents have changed.

👍 for those who do not use auto-refresh

![image](https://github.com/user-attachments/assets/4a038ec4-eb26-4ec0-be1f-5b91c1b7b70e)

I am going to add a split button with this setting, too.

GitStatusMonitor status to be added in a follow-up
BTW: Even with f0b5a24b3997c86b4fd23c99433d3bb4148206ee, GitStatusMonitor updates can be pretty slow - particularly with an AV virus.

gerhardol commented:
> > When is Dense tree/Group nodes not disabled? Is dropdown needed?
> 
>     * not implemented

Remove from PR?

> > Noted, probably OK:
> > If a folder does not exist on disk "Show in folder" is disabled, as files are. Good. If top level group or status/extension status is shown it is available, opening the repo. Maybe change text to "Show repo folder"?
> 
> I do not like to exchange menu item texts on runtime unless really necessary. I think opening the repo folder for root / group-by nodes is not counter-intuitive.

It was not what I expected. No objection for now.

> > (maybe not this PR) If selecting a folder, "Save as" and others are available. This should maybe select the files in the folder too.
> 
> If you mean "Show in folder": A folder is selected. So I think, that this folder shall be selected in the Explorer. Selecting multiple files opens multiple Explorer windows. If users really want all files to be shown in Explorer (windows), they should explicitly use "Select all" first.

OK

> > Add File tree group, similar to the grep group. The file tree is improved in an update, but then it would not be needed.
> 
> File tree needs slightly different handling. E.g. grouping does not really apply. I prefer to not add more groups, especially no expensive ones. It would need to be excluded from "find text in files".

This group should only be calculated when expanded, we leave it here for now.

> > Refresh button, for artificial commits. That can later show the status from GitStatusMonitor if the contents have changed.
> 
> 👍 for those who do not use auto-refresh

Auto refresh is too slow for me in some repos

> I am going to add a split button with this setting, too.
> 
> GitStatusMonitor status to be added in a follow-up BTW: Even with [f0b5a24](https://github.com/gitextensions/gitextensions/commit/f0b5a24b3997c86b4fd23c99433d3bb4148206ee), GitStatusMonitor updates can be pretty slow - particularly with an AV virus.

I would have preferred to have it leftmost as for the main toolbar and being disabled if no artificial. Now it is moving places...

mstv commented:
> leftmost as for the main toolbar and being disabled if no artificial. Now it is moving places...

static position now, but the Collapse button is more important

![image](https://github.com/user-attachments/assets/398bd35f-ed44-41c5-b2a3-e0fec0187d62)

> can **later** show the status from GitStatusMonitor if the contents have changed

Currently, this essential information is not displayed (or not retrieved). The numbers at the commit button and at the artificial commits do not really help in indicating whether a refresh is needed or not.
This toolbar must not contain text. Text is very distracting here. Dynamic texts would be even worse.
![image](https://github.com/user-attachments/assets/3f1e92e6-9c26-4059-80fc-1bc3e0823846) "ReloadRevisionsDirty.png" should be used in order to indicate that a refresh is required.

gerhardol commented:
> Currently, this essential information is not displayed (or not retrieved). The numbers at the commit button and at the artificial commits do not really help in indicating whether a refresh is needed or not.

It would have to compare current status with the update, if ever implemented.

> This toolbar must not contain text. Text is very distracting here. Dynamic texts would be even worse.

Agree, numbers are in the group (may be different) and grid.

gerhardol commented on src/app/GitUI/UserControls/FileStatusList.Toolbar.cs:
```diff
@@ -0,0 +1,196 @@
+
+    private void UpdateToolbar(IReadOnlyList<GitRevision> revisions)
+    {
+        btnRefresh.Enabled = revisions.Any(revision => revision.IsArtificial);
```
When filtering files in a certain (private) repo (also when refreshing) in filetree tab
revisions can seemingly be null. Check here or is the caller incorrect?

```
System.NullReferenceException: Object reference not set to an instance of an object.
   at void GitUI.FileStatusList.UpdateToolbar()+(GitRevision revision) => { } in C:/dev/gc/gitextensions/src/app/GitUI/UserControls/FileStatusList.Toolbar.cs:line 194
   at bool System.Linq.Enumerable.Any<TSource>(IEnumerable<TSource> source, Func<TSource, bool> predicate)
   at void GitUI.FileStatusList.UpdateToolbar(IReadOnlyList<GitRevision> revisions) in C:/dev/gc/gitextensions/src/app/GitUI/UserControls/FileStatusList.Toolbar.cs:line 194
   at async Task GitUI.FileStatusList.SetDiffsAsync(IReadOnlyList<GitRevision> revisions, ObjectId headId, CancellationToken cancellationToken) in C:/dev/gc/gitextensions/src/app/GitUI/UserControls/FileStatusList.cs:line 865
   at async Task GitUI.CommandsDialogs.RevisionDiffControl.SetDiffsAsync(IReadOnlyList<GitRevision> revisions) in C:/dev/gc/gitextensions/src/app/GitUI/CommandsDialogs/RevisionDiffControl.cs:line 341
   at async void GitUI.CommandsDialogs.RevisionDiffControl.DisplayDiffTab(IReadOnlyList<GitRevision> revisions)+(?) => { } in C:/dev/gc/gitextensions/src/app/GitUI/CommandsDialogs/RevisionDiffControl.cs:line 273
   at async void GitUI.TaskManager.InvokeAndForget(Control control, Func<Task> asyncAction, CancellationToken cancellationToken)+(?) => { } [1] in C:/dev/gc/gitextensions/src/app/GitExtUtils/GitUI/TaskManager.cs:line 119
   at async Task GitUI.TaskManager.HandleExceptionsAsync(Func<Task> asyncAction, Func<Exception, Task> handleExceptionAsync) in C:/dev/gc/gitextensions/src/app/GitExtUtils/GitUI/TaskManager.cs:line 32
```

Using build based on your branch with commits on top.

mstv replied:
Thank you for the callstack. `nullable` declaration was missing in `FormBrowse.FillFileTree`.

gerhardol reviewed (commented):
+1 no detailed review yet but it works well

gerhardol commented on src/app/GitUI/CommandsDialogs/RevisionDiffControl.cs:
```diff
@@ -620,9 +620,42 @@ await DiffText.ViewChangesAsync(DiffFiles.SelectedItem,
+                }
+
+                StringBuilder description = new();
+                description.Append('(').Append(items.Length).Append(") ").AppendLine(path);
```
?
```suggestion
                description.Append($"({items.Length}) {path}");
```

mstv replied:
This may be shorter and easier to read. But there is a StringBuilder already, and I wanted to avoid building an interpolated string in addition.

gerhardol reviewed (commented):
works well
I have been using this for the >1 month it has been public. Review has been done over time, this final review was not so deep but I find this to be a code improvement

gerhardol commented:
Related, not this PR:
* This prepares for FileStatusList in FileTree. Submodules are not displayed in filetree any more.

gerhardol commented:
What about this PR?
I would like to get forward with dark mode, next is #12167 etc before #12064 and I consider the FileStatus changes to precede this. Then #12111 can be reviewed
Can I merge?

mstv commented on src/app/GitUI/UserControls/FileStatusList.cs:
```diff
@@ -292,8 +294,12 @@ void UICommands_PostSettings(object? sender, GitUIPostActionEventArgs? e)
         }
 
-        public void Bind(Func<ObjectId?, string> describeRevision, Func<GitRevision, GitRevision> getActualRevision)
+        public void Bind(Action refreshArtificial, bool canAutoRefresh = false, Func<ObjectId?, string>? describeRevision = null, Func<GitRevision, GitRevision>? getActualRevision = null)
```
It should not be a big deal to use the refresh button of the new toolbar in FormCommit, too - in a follow-up.

RussKie commented on src/app/GitUI/CommandsDialogs/FormStash.cs:
```diff
@@ -286,18 +286,17 @@ private void LoadGitItemStatuses(IReadOnlyList<GitItemStatus> gitItemStatuses)
         private void ResizeStashesWidth()
         {
-            Stashes.Size = new Size(toolStrip1.Width - 15 - refreshToolStripButton.Width - showToolStripLabel.Width, Stashes.Size.Height);
+            Stashes.Width = toolStrip1.Width - 5 - showToolStripLabel.Width;
```
It would great to document what "5" is.

mstv replied:
Adding a `const`. I had just removed the `refreshToolStripButton` and its spacing.
Please do not blame me for the layout of the rather old FormStash nor for limitations of widgets.
ToolStripItems do not have neither alignment "fill" nor public properties `Left` and `Right`.

RussKie commented on src/app/GitUI/CommandsDialogs/RevisionDiffController.cs:
```diff
@@ -27,67 +27,27 @@ public interface IRevisionDiffController
     }
 
-    public sealed class ContextMenuSelectionInfo
+    internal sealed record ContextMenuSelectionInfo(
```
Please move out into own file.
RevisionDiffContoller.ContextMenuSelectionInfo.cs I guess?

gerhardol commented:
5d8600673dc07d5a43f4489112f20780d00a6c71

![image](https://github.com/user-attachments/assets/4ce2ad78-ee06-4d86-bd5c-73c3144e7ee7)

Also, with mstv/feature/filetree
No info for renamed cars (was OK earlier)
![image](https://github.com/user-attachments/assets/670f353e-ac91-40cd-bb93-dc1a88760499)

mdonatas commented:
Sorry in advance @mstv if this PR is not responsible for this behavior but since you tagged me here I tested your `_all` branch and there is a jiggle in certain situation upon root node click.

Also, the icons are not scaled.

https://github.com/user-attachments/assets/6ce230c1-95ca-4f98-8708-ebe6d15cca1a

mdonatas commented:
There probably should be a separator here to better indicate a "radio button group". These look exactly like the icons above the left panel or the panel toggle buttons in the main form's toolbar but they act differently, e.g. nothing happens if you press an already active icon. I am not sure if there is a better solution... a bit not feng-shui that same looks have different behavior (toggle button vs radio button group) but probably not a deal-breaker

![image](https://github.com/user-attachments/assets/c6c798eb-2419-4a27-a73e-e9aa2dab830c)

mdonatas commented:
shift-click to select multiple items is broken. It only works if upon mouse-down on a second item the mouse is slightly moved before mouse-up. But when a second items is simply clicked without moving the mouse during click, selection doesn't happen.

mdonatas commented:
It took me quite some time to figure out when "Toggle flat list / tree" options in combination with grouping-options had what effect. Have you considered to have two radio-button like behaving toggle buttons like it is with grouping-options? Then each of them could have settings related to it in a drop-down. There is a good chance that current drop-down would gain more settings in the future where they too would be relevant to one of the modes only, there is a good chance for this to become even messier.

![image](https://github.com/user-attachments/assets/c0c9ccb2-9cba-4dd2-8c81-4e24f646732e)

mstv commented:
Thank you for the feedback!
The last, - I admit - quickly added workaround for the workaround broke things.
Here comes the workaround for the workaround for workaround... It repairs Shift+Click, Drag & Drop.
I have even found a workaround for the jiggle = wiggle - at least for the major occasions. If anything remains, it is the TreeView living its own life.

Icon scaling should be the only open issue. I want to handle it separately.

> There probably should be a separator here to better indicate a "radio button group".

I had already added a separator and removed it because the drop arrow seemed and seems to be sufficiently separating this group.

mdonatas commented:
Activate on mouse-down is so much better UX wise, the muscle-memory of how I am used to GE working is very strong. Thank you for this effort!

---

Not suggesting to make these changes now, but what do you think if there was a ⚙️ icon in the toolbar which would open a settings dialog for related options? It would remove the need to have a down-arrow on the "Toggle flat list / tree" button and have you considered to have two buttons rather than one?

Toolbar arrangement could be `Show-Tree Show-Flat | Group-By-A Group-By-B Group-By-C | ⚙️`

mstv commented:
> but what do you think if there was a ⚙️ icon in the toolbar which would open a settings dialog for related options?

No more additional settings pages, please. And settings pages are slow.

> It would remove the need to have a down-arrow on the "Toggle flat list / tree" button and have you considered to have two buttons rather than one?

The drop-down menu contains settings which are changed least frequently - but they are pretty related to this toolbar button. The result can directly be seen.
Additional buttons consume more space for no gain. The blue highlighting of the active group-by option is already distracting enough.

gerhardol commented:
> I think there's enough space to accommodate the required buttons. I also think, the treeview/listview switch should be a single button. My 2c

There are more buttons when comparing branches, so it is getting full
![image](https://github.com/user-attachments/assets/e42cd11c-0097-4cec-97f0-5a137917a58d)

I find the current status to be good

gerhardol commented:
As this PR improves on the FileStatusList and there are other threads reporting on issues, is this ready to be merged?
I find this to be working well (but also saw the improvements in #12116 over the in my view small issues).

mstv commented:
There were a lot of fixes needed to this PR and the previous PR, I admit. All known issues should have been resolved.

RussKie commented:
![image](https://github.com/user-attachments/assets/76917a9a-3d0c-407b-b210-5f6145f871e9)

The last 3 buttons appear to be mutually-exclusive, and I struggled to get the buttons purpose by looking at the icons. I don't think this is a desired UX, and there's a room for improvement.
I suggest looking at how we implement sorting as well as looking at how Windows Explorer implements grouping - the concept is the same - there's one dropdown menu/button that shows available options:

![image](https://github.com/user-attachments/assets/8f0ef1bf-37d5-4dea-bb9e-1453d3ea766e)
![image](https://github.com/user-attachments/assets/b5423eb8-04f7-4a2f-a17c-3cb441c6ae2e)

mstv commented:
![grafik](https://github.com/user-attachments/assets/2917a976-2546-4dc9-b55d-cbadfc60253c)

![grafik](https://github.com/user-attachments/assets/95e19ed6-b14d-49c4-a5bf-d21ef0494f06)

mstv commented:
This toolbar must not contain text. It would be too distracting.

RussKie commented:
It already is, if I'm being honest. I keep clicking on minimise button instead of the refresh.
I don't think it'll get any more distracting, but it'll possibly make it more intuitive for users. 

It's hard to tell exactly until we see it; it happened on many occasions before where we discussed it for weeks just to find out that an agreed UI just didn't work, and the dismissed alternatives were later be deemed the best option.

mstv commented:
I have added more separators and yet another place for "group by" items.
You can hide toolbar items now (not persisted yet):

![image](https://github.com/user-attachments/assets/f358414b-754c-4947-b899-6ab326f3a84f)

This is and shall be a toolbar with text in (existing!) tooltips only (but not a ribbon with drop-downs).
Switching modes by direct click is crucial.

mdonatas commented:
There is a slight bug with the 🔎toggle when it's set to `Both` when the dialog is closed via its ❌ button rather than pressing 🔎 again.
Basically after closing the dialog and pressing 🔎 again to reveal it would not do that but rather hide the regex search box, so we have 3 states
- Nothing
- Dialgo and regex box
- Regex box (this state should not be available when `Both` is selected)

https://github.com/user-attachments/assets/7a64b1c2-7d52-42ab-8bc3-98e0c68620f2

mdonatas commented:
One more bug. "Show files changed..." filter states are not applied when repeating a diff. Notice when I do the same diff again the filter's state doesn't match the contents.

https://github.com/user-attachments/assets/8533ce56-641e-4ede-be4f-0f4ef2d63333

mdonatas commented:
Mismatched text. From clicking around these seem to be the same buttons in different places yet their text is different: _sort_ VS _group_

![image](https://github.com/user-attachments/assets/7696b43b-4611-4f7d-a68c-dba08cbfd1fe)

![image](https://github.com/user-attachments/assets/a5a7d21b-bf05-4e02-8bf0-ea6ff33ce83a)

mstv commented:
Good catches, @mdonatas!

> 3 states: Nothing / Dialog and regex box / Regex box (this state should not be available when Both is selected)

I have improved the behavior for the third (and forth) state: hide only if both are visible else show both.
(Actually, the fourth state is possible, too.)

---

> "Sort" vs. "Group by"

I have added "and group by" to the old context menu item. I have kept "&Sort" for the mnemonic.
(I do not want to add "Sort and" to the toolbar. I would rather remove the context menu item.)

---

> Filter not re-applied

I have removed a shortcut (too late update of the toolbar button's visibility).

mstv commented:
Sorry, yet another two addenda

mdonatas commented:
Huge effort, congratulations! 🚀
